// @wordpress/widgets/build/blocks/widget-group/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Edit;
var _react = require("react");
var _blockEditor = require("@wordpress/block-editor");
var _components = require("@wordpress/components");
var _icons = require("@wordpress/icons");
var _i18n = require("@wordpress/i18n");
var _data = require("@wordpress/data");
/**
 * WordPress dependencies
 */

function Edit(props) {
  const {
    clientId
  } = props;
  const {
    innerBlocks
  } = (0, _data.useSelect)(select => select(_blockEditor.store).getBlock(clientId), [clientId]);
  return (0, _react.createElement)("div", {
    ...(0, _blockEditor.useBlockProps)({
      className: 'widget'
    })
  }, innerBlocks.length === 0 ? (0, _react.createElement)(PlaceholderContent, {
    ...props
  }) : (0, _react.createElement)(PreviewContent, {
    ...props
  }));
}
function PlaceholderContent({
  clientId
}) {
  return (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_components.Placeholder, {
    className: "wp-block-widget-group__placeholder",
    icon: (0, _react.createElement)(_blockEditor.BlockIcon, {
      icon: _icons.group
    }),
    label: (0, _i18n.__)('Widget Group')
  }, (0, _react.createElement)(_blockEditor.ButtonBlockAppender, {
    rootClientId: clientId
  })), (0, _react.createElement)(_blockEditor.InnerBlocks, {
    renderAppender: false
  }));
}
function PreviewContent({
  attributes,
  setAttributes
}) {
  var _attributes$title;
  return (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_blockEditor.RichText, {
    tagName: "h2",
    identifier: "title",
    className: "widget-title",
    allowedFormats: [],
    placeholder: (0, _i18n.__)('Title'),
    value: (_attributes$title = attributes.title) !== null && _attributes$title !== void 0 ? _attributes$title : '',
    onChange: title => setAttributes({
      title
    })
  }), (0, _react.createElement)(_blockEditor.InnerBlocks, null));
}
