// @wordpress/edit-post/build/components/header/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _editor = require("@wordpress/editor");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _fullscreenModeClose = _interopRequireDefault(require("./fullscreen-mode-close"));
var _moreMenu = _interopRequireDefault(require("./more-menu"));
var _mainDashboardButton = _interopRequireDefault(require("./main-dashboard-button"));
var _store = require("../../store");
var _lockUnlock = require("../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Header: EditorHeader
} = (0, _lockUnlock.unlock)(_editor.privateApis);
const slideX = {
  hidden: {
    x: '-100%'
  },
  distractionFreeInactive: {
    x: 0
  },
  hover: {
    x: 0,
    transition: {
      type: 'tween',
      delay: 0.2
    }
  }
};
function Header({
  setEntitiesSavedStatesCallback,
  initialPost
}) {
  const {
    hasActiveMetaboxes
  } = (0, _data.useSelect)(select => {
    return {
      hasActiveMetaboxes: select(_store.store).hasMetaBoxes()
    };
  }, []);
  return (0, _react.createElement)(EditorHeader, {
    forceIsDirty: hasActiveMetaboxes,
    setEntitiesSavedStatesCallback: setEntitiesSavedStatesCallback
  }, (0, _react.createElement)(_mainDashboardButton.default.Slot, null, (0, _react.createElement)(_components.__unstableMotion.div, {
    variants: slideX,
    transition: {
      type: 'tween',
      delay: 0.8
    }
  }, (0, _react.createElement)(_fullscreenModeClose.default, {
    showTooltip: true,
    initialPost: initialPost
  }))), (0, _react.createElement)(_moreMenu.default, null));
}
var _default = exports.default = Header;
