// @wordpress/edit-post/build/components/text-editor/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TextEditor;
var _react = require("react");
var _editor = require("@wordpress/editor");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _i18n = require("@wordpress/i18n");
var _keycodes = require("@wordpress/keycodes");
var _element = require("@wordpress/element");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function TextEditor() {
  const isRichEditingEnabled = (0, _data.useSelect)(select => {
    return select(_editor.store).getEditorSettings().richEditingEnabled;
  }, []);
  const {
    switchEditorMode
  } = (0, _data.useDispatch)(_editor.store);
  const {
    isWelcomeGuideVisible
  } = (0, _data.useSelect)(select => {
    const {
      isFeatureActive
    } = select(_store.store);
    return {
      isWelcomeGuideVisible: isFeatureActive('welcomeGuide')
    };
  }, []);
  const titleRef = (0, _element.useRef)();
  (0, _element.useEffect)(() => {
    if (isWelcomeGuideVisible) {
      return;
    }
    titleRef?.current?.focus();
  }, [isWelcomeGuideVisible]);
  return (0, _react.createElement)("div", {
    className: "edit-post-text-editor"
  }, isRichEditingEnabled && (0, _react.createElement)("div", {
    className: "edit-post-text-editor__toolbar"
  }, (0, _react.createElement)("h2", null, (0, _i18n.__)('Editing code')), (0, _react.createElement)(_components.Button, {
    variant: "tertiary",
    onClick: () => switchEditorMode('visual'),
    shortcut: _keycodes.displayShortcut.secondary('m')
  }, (0, _i18n.__)('Exit code editor'))), (0, _react.createElement)("div", {
    className: "edit-post-text-editor__body"
  }, (0, _react.createElement)(_editor.PostTitleRaw, {
    ref: titleRef
  }), (0, _react.createElement)(_editor.PostTextEditor, null)));
}
