// @wordpress/widgets/build/components/move-to-widget-area/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MoveToWidgetArea;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
/**
 * WordPress dependencies
 */

function MoveToWidgetArea({
  currentWidgetAreaId,
  widgetAreas,
  onSelect
}) {
  return (0, _react.createElement)(_components.ToolbarGroup, null, (0, _react.createElement)(_components.ToolbarItem, null, toggleProps => (0, _react.createElement)(_components.DropdownMenu, {
    icon: _icons.moveTo,
    label: (0, _i18n.__)('Move to widget area'),
    toggleProps: toggleProps
  }, ({
    onClose
  }) => (0, _react.createElement)(_components.MenuGroup, {
    label: (0, _i18n.__)('Move to')
  }, (0, _react.createElement)(_components.MenuItemsChoice, {
    choices: widgetAreas.map(widgetArea => ({
      value: widgetArea.id,
      label: widgetArea.name,
      info: widgetArea.description
    })),
    value: currentWidgetAreaId,
    onSelect: value => {
      onSelect(value);
      onClose();
    }
  })))));
}
