// @wordpress/editor/build/components/post-publish-panel/maybe-post-format-panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostFormatPanel;
var _react = require("react");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _i18n = require("@wordpress/i18n");
var _coreData = require("@wordpress/core-data");
var _postFormat = require("../post-format");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const getSuggestion = (supportedFormats, suggestedPostFormat) => {
  const formats = _postFormat.POST_FORMATS.filter(format => supportedFormats?.includes(format.id));
  return formats.find(format => format.id === suggestedPostFormat);
};
const PostFormatSuggestion = ({
  suggestedPostFormat,
  suggestionText,
  onUpdatePostFormat
}) => (0, _react.createElement)(_components.Button, {
  variant: "link",
  onClick: () => onUpdatePostFormat(suggestedPostFormat)
}, suggestionText);
function PostFormatPanel() {
  const {
    currentPostFormat,
    suggestion
  } = (0, _data.useSelect)(select => {
    var _select$getThemeSuppo;
    const {
      getEditedPostAttribute,
      getSuggestedPostFormat
    } = select(_store.store);
    const supportedFormats = (_select$getThemeSuppo = select(_coreData.store).getThemeSupports().formats) !== null && _select$getThemeSuppo !== void 0 ? _select$getThemeSuppo : [];
    return {
      currentPostFormat: getEditedPostAttribute('format'),
      suggestion: getSuggestion(supportedFormats, getSuggestedPostFormat())
    };
  }, []);
  const {
    editPost
  } = (0, _data.useDispatch)(_store.store);
  const onUpdatePostFormat = format => editPost({
    format
  });
  const panelBodyTitle = [(0, _i18n.__)('Suggestion:'), (0, _react.createElement)("span", {
    className: "editor-post-publish-panel__link",
    key: "label"
  }, (0, _i18n.__)('Use a post format'))];
  if (!suggestion || suggestion.id === currentPostFormat) {
    return null;
  }
  return (0, _react.createElement)(_components.PanelBody, {
    initialOpen: false,
    title: panelBodyTitle
  }, (0, _react.createElement)("p", null, (0, _i18n.__)('Your theme uses post formats to highlight different kinds of content, like images or videos. Apply a post format to see this special styling.')), (0, _react.createElement)("p", null, (0, _react.createElement)(PostFormatSuggestion, {
    onUpdatePostFormat: onUpdatePostFormat,
    suggestedPostFormat: suggestion.id,
    suggestionText: (0, _i18n.sprintf)( /* translators: %s: post format */
    (0, _i18n.__)('Apply the "%1$s" format.'), suggestion.caption)
  })));
}
