// @wordpress/editor/build/components/post-publish-panel/maybe-tags-panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _coreData = require("@wordpress/core-data");
var _flatTermSelector = _interopRequireDefault(require("../post-taxonomies/flat-term-selector"));
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const TagsPanel = () => {
  const panelBodyTitle = [(0, _i18n.__)('Suggestion:'), (0, _react.createElement)("span", {
    className: "editor-post-publish-panel__link",
    key: "label"
  }, (0, _i18n.__)('Add tags'))];
  return (0, _react.createElement)(_components.PanelBody, {
    initialOpen: false,
    title: panelBodyTitle
  }, (0, _react.createElement)("p", null, (0, _i18n.__)('Tags help users and search engines navigate your site and find your content. Add a few keywords to describe your post.')), (0, _react.createElement)(_flatTermSelector.default, {
    slug: 'post_tag'
  }));
};
const MaybeTagsPanel = () => {
  const {
    hasTags,
    isPostTypeSupported
  } = (0, _data.useSelect)(select => {
    const postType = select(_store.store).getCurrentPostType();
    const tagsTaxonomy = select(_coreData.store).getTaxonomy('post_tag');
    const _isPostTypeSupported = tagsTaxonomy?.types?.includes(postType);
    const areTagsFetched = tagsTaxonomy !== undefined;
    const tags = tagsTaxonomy && select(_store.store).getEditedPostAttribute(tagsTaxonomy.rest_base);
    return {
      hasTags: !!tags?.length,
      isPostTypeSupported: areTagsFetched && _isPostTypeSupported
    };
  }, []);
  const [hadTagsWhenOpeningThePanel] = (0, _element.useState)(hasTags);
  if (!isPostTypeSupported) {
    return null;
  }

  /*
   * We only want to show the tag panel if the post didn't have
   * any tags when the user hit the Publish button.
   *
   * We can't use the prop.hasTags because it'll change to true
   * if the user adds a new tag within the pre-publish panel.
   * This would force a re-render and a new prop.hasTags check,
   * hiding this panel and keeping the user from adding
   * more than one tag.
   */
  if (!hadTagsWhenOpeningThePanel) {
    return (0, _react.createElement)(TagsPanel, null);
  }
  return null;
};
var _default = exports.default = MaybeTagsPanel;
