// @wordpress/editor/build/components/post-featured-image/panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostFeaturedImagePanel;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _coreData = require("@wordpress/core-data");
var _store = require("../../store");
var _index = _interopRequireDefault(require("./index"));
var _check = _interopRequireDefault(require("./check"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const PANEL_NAME = 'featured-image';
function PostFeaturedImagePanel({
  withPanelBody = true
}) {
  var _postType$labels$feat;
  const {
    postType,
    isEnabled,
    isOpened
  } = (0, _data.useSelect)(select => {
    const {
      getEditedPostAttribute,
      isEditorPanelEnabled,
      isEditorPanelOpened
    } = select(_store.store);
    const {
      getPostType
    } = select(_coreData.store);
    return {
      postType: getPostType(getEditedPostAttribute('type')),
      isEnabled: isEditorPanelEnabled(PANEL_NAME),
      isOpened: isEditorPanelOpened(PANEL_NAME)
    };
  }, []);
  const {
    toggleEditorPanelOpened
  } = (0, _data.useDispatch)(_store.store);
  if (!isEnabled) {
    return null;
  }
  if (!withPanelBody) {
    return (0, _react.createElement)(_check.default, null, (0, _react.createElement)(_index.default, null));
  }
  return (0, _react.createElement)(_check.default, null, (0, _react.createElement)(_components.PanelBody, {
    title: (_postType$labels$feat = postType?.labels?.featured_image) !== null && _postType$labels$feat !== void 0 ? _postType$labels$feat : (0, _i18n.__)('Featured image'),
    opened: isOpened,
    onToggle: () => toggleEditorPanelOpened(PANEL_NAME)
  }, (0, _react.createElement)(_index.default, null)));
}
