// @wordpress/interface/build/components/complementary-area-more-menu-item/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ComplementaryAreaMoreMenuItem;
var _react = require("react");
var _icons = require("@wordpress/icons");
var _components = require("@wordpress/components");
var _complementaryAreaToggle = _interopRequireDefault(require("../complementary-area-toggle"));
var _actionItem = _interopRequireDefault(require("../action-item"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const PluginsMenuItem = ({
  // Menu item is marked with unstable prop for backward compatibility.
  // They are removed so they don't leak to DOM elements.
  // @see https://github.com/WordPress/gutenberg/issues/14457
  __unstableExplicitMenuItem,
  __unstableTarget,
  ...restProps
}) => (0, _react.createElement)(_components.MenuItem, {
  ...restProps
});
function ComplementaryAreaMoreMenuItem({
  scope,
  target,
  __unstableExplicitMenuItem,
  ...props
}) {
  return (0, _react.createElement)(_complementaryAreaToggle.default, {
    as: toggleProps => {
      return (0, _react.createElement)(_actionItem.default, {
        __unstableExplicitMenuItem: __unstableExplicitMenuItem,
        __unstableTarget: `${scope}/${target}`,
        as: PluginsMenuItem,
        name: `${scope}/plugin-more-menu`,
        ...toggleProps
      });
    },
    role: "menuitemcheckbox",
    selectedIcon: _icons.check,
    name: target,
    scope: scope,
    ...props
  });
}
