// @wordpress/router/build/router.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RouterProvider = RouterProvider;
exports.useHistory = useHistory;
exports.useLocation = useLocation;
var _react = require("react");
var _element = require("@wordpress/element");
var _history = _interopRequireDefault(require("./history"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const RoutesContext = (0, _element.createContext)();
const HistoryContext = (0, _element.createContext)();
function useLocation() {
  return (0, _element.useContext)(RoutesContext);
}
function useHistory() {
  return (0, _element.useContext)(HistoryContext);
}
function getLocationWithParams(location) {
  const searchParams = new URLSearchParams(location.search);
  return {
    ...location,
    params: Object.fromEntries(searchParams.entries())
  };
}
function RouterProvider({
  children
}) {
  const [location, setLocation] = (0, _element.useState)(() => getLocationWithParams(_history.default.location));
  (0, _element.useEffect)(() => {
    return _history.default.listen(({
      location: updatedLocation
    }) => {
      setLocation(getLocationWithParams(updatedLocation));
    });
  }, []);
  return (0, _react.createElement)(HistoryContext.Provider, {
    value: _history.default
  }, (0, _react.createElement)(RoutesContext.Provider, {
    value: location
  }, children));
}
