// @wordpress/block-library/build/navigation-link/link-ui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkUI = LinkUI;
exports.default = void 0;
exports.getSuggestionsQuery = getSuggestionsQuery;
var _react = require("react");
var _dom = require("@wordpress/dom");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _blockEditor = require("@wordpress/block-editor");
var _element = require("@wordpress/element");
var _coreData = require("@wordpress/core-data");
var _htmlEntities = require("@wordpress/html-entities");
var _data = require("@wordpress/data");
var _icons = require("@wordpress/icons");
var _compose = require("@wordpress/compose");
var _lockUnlock = require("../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  PrivateQuickInserter: QuickInserter
} = (0, _lockUnlock.unlock)(_blockEditor.privateApis);

/**
 * Given the Link block's type attribute, return the query params to give to
 * /wp/v2/search.
 *
 * @param {string} type Link block's type attribute.
 * @param {string} kind Link block's entity of kind (post-type|taxonomy)
 * @return {{ type?: string, subtype?: string }} Search query params.
 */
function getSuggestionsQuery(type, kind) {
  switch (type) {
    case 'post':
    case 'page':
      return {
        type: 'post',
        subtype: type
      };
    case 'category':
      return {
        type: 'term',
        subtype: 'category'
      };
    case 'tag':
      return {
        type: 'term',
        subtype: 'post_tag'
      };
    case 'post_format':
      return {
        type: 'post-format'
      };
    default:
      if (kind === 'taxonomy') {
        return {
          type: 'term',
          subtype: type
        };
      }
      if (kind === 'post-type') {
        return {
          type: 'post',
          subtype: type
        };
      }
      return {
        // for custom link which has no type
        // always show pages as initial suggestions
        initialSuggestionsSearchOptions: {
          type: 'post',
          subtype: 'page',
          perPage: 20
        }
      };
  }
}
function LinkUIBlockInserter({
  clientId,
  onBack,
  onSelectBlock
}) {
  const {
    rootBlockClientId
  } = (0, _data.useSelect)(select => {
    const {
      getBlockRootClientId
    } = select(_blockEditor.store);
    return {
      rootBlockClientId: getBlockRootClientId(clientId)
    };
  }, [clientId]);
  const focusOnMountRef = (0, _compose.useFocusOnMount)('firstElement');
  const dialogTitleId = (0, _compose.useInstanceId)(_blockEditor.__experimentalLinkControl, `link-ui-block-inserter__title`);
  const dialogDescritionId = (0, _compose.useInstanceId)(_blockEditor.__experimentalLinkControl, `link-ui-block-inserter__description`);
  if (!clientId) {
    return null;
  }
  return (0, _react.createElement)("div", {
    className: "link-ui-block-inserter",
    role: "dialog",
    "aria-labelledby": dialogTitleId,
    "aria-describedby": dialogDescritionId,
    ref: focusOnMountRef
  }, (0, _react.createElement)(_components.VisuallyHidden, null, (0, _react.createElement)("h2", {
    id: dialogTitleId
  }, (0, _i18n.__)('Add block')), (0, _react.createElement)("p", {
    id: dialogDescritionId
  }, (0, _i18n.__)('Choose a block to add to your Navigation.'))), (0, _react.createElement)(_components.Button, {
    className: "link-ui-block-inserter__back",
    icon: _icons.chevronLeftSmall,
    onClick: e => {
      e.preventDefault();
      onBack();
    },
    size: "small"
  }, (0, _i18n.__)('Back')), (0, _react.createElement)(QuickInserter, {
    rootClientId: rootBlockClientId,
    clientId: clientId,
    isAppender: false,
    prioritizePatterns: false,
    selectBlockOnInsert: true,
    hasSearch: false,
    onSelect: onSelectBlock
  }));
}
function LinkUI(props) {
  const [addingBlock, setAddingBlock] = (0, _element.useState)(false);
  const [focusAddBlockButton, setFocusAddBlockButton] = (0, _element.useState)(false);
  const {
    saveEntityRecord
  } = (0, _data.useDispatch)(_coreData.store);
  const pagesPermissions = (0, _coreData.useResourcePermissions)('pages');
  const postsPermissions = (0, _coreData.useResourcePermissions)('posts');
  async function handleCreate(pageTitle) {
    const postType = props.link.type || 'page';
    const page = await saveEntityRecord('postType', postType, {
      title: pageTitle,
      status: 'draft'
    });
    return {
      id: page.id,
      type: postType,
      // Make `title` property consistent with that in `fetchLinkSuggestions` where the `rendered` title (containing HTML entities)
      // is also being decoded. By being consistent in both locations we avoid having to branch in the rendering output code.
      // Ideally in the future we will update both APIs to utilise the "raw" form of the title which is better suited to edit contexts.
      // e.g.
      // - title.raw = "Yes & No"
      // - title.rendered = "Yes &#038; No"
      // - decodeEntities( title.rendered ) = "Yes & No"
      // See:
      // - https://github.com/WordPress/gutenberg/pull/41063
      // - https://github.com/WordPress/gutenberg/blob/a1e1fdc0e6278457e9f4fc0b31ac6d2095f5450b/packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js#L212-L218
      title: (0, _htmlEntities.decodeEntities)(page.title.rendered),
      url: page.link,
      kind: 'post-type'
    };
  }
  const {
    label,
    url,
    opensInNewTab,
    type,
    kind
  } = props.link;
  let userCanCreate = false;
  if (!type || type === 'page') {
    userCanCreate = pagesPermissions.canCreate;
  } else if (type === 'post') {
    userCanCreate = postsPermissions.canCreate;
  }

  // Memoize link value to avoid overriding the LinkControl's internal state.
  // This is a temporary fix. See https://github.com/WordPress/gutenberg/issues/50976#issuecomment-1568226407.
  const link = (0, _element.useMemo)(() => ({
    url,
    opensInNewTab,
    title: label && (0, _dom.__unstableStripHTML)(label)
  }), [label, opensInNewTab, url]);
  const dialogTitleId = (0, _compose.useInstanceId)(LinkUI, `link-ui-link-control__title`);
  const dialogDescritionId = (0, _compose.useInstanceId)(LinkUI, `link-ui-link-control__description`);

  // Selecting a block should close the popover and also remove the (previously) automatically inserted
  // link block so that the newly selected block can be inserted in its place.
  const {
    onClose: onSelectBlock
  } = props;
  return (0, _react.createElement)(_components.Popover, {
    placement: "bottom",
    onClose: props.onClose,
    anchor: props.anchor,
    shift: true
  }, !addingBlock && (0, _react.createElement)("div", {
    role: "dialog",
    "aria-labelledby": dialogTitleId,
    "aria-describedby": dialogDescritionId
  }, (0, _react.createElement)(_components.VisuallyHidden, null, (0, _react.createElement)("h2", {
    id: dialogTitleId
  }, (0, _i18n.__)('Add link')), (0, _react.createElement)("p", {
    id: dialogDescritionId
  }, (0, _i18n.__)('Search for and add a link to your Navigation.'))), (0, _react.createElement)(_blockEditor.__experimentalLinkControl, {
    hasTextControl: true,
    hasRichPreviews: true,
    value: link,
    showInitialSuggestions: true,
    withCreateSuggestion: userCanCreate,
    createSuggestion: handleCreate,
    createSuggestionButtonText: searchTerm => {
      let format;
      if (type === 'post') {
        /* translators: %s: search term. */
        format = (0, _i18n.__)('Create draft post: <mark>%s</mark>');
      } else {
        /* translators: %s: search term. */
        format = (0, _i18n.__)('Create draft page: <mark>%s</mark>');
      }
      return (0, _element.createInterpolateElement)((0, _i18n.sprintf)(format, searchTerm), {
        mark: (0, _react.createElement)("mark", null)
      });
    },
    noDirectEntry: !!type,
    noURLSuggestion: !!type,
    suggestionsQuery: getSuggestionsQuery(type, kind),
    onChange: props.onChange,
    onRemove: props.onRemove,
    onCancel: props.onCancel,
    renderControlBottom: () => !link?.url?.length && (0, _react.createElement)(LinkUITools, {
      focusAddBlockButton: focusAddBlockButton,
      setAddingBlock: () => {
        setAddingBlock(true);
        setFocusAddBlockButton(false);
      }
    })
  })), addingBlock && (0, _react.createElement)(LinkUIBlockInserter, {
    clientId: props.clientId,
    onBack: () => {
      setAddingBlock(false);
      setFocusAddBlockButton(true);
    },
    onSelectBlock: onSelectBlock
  }));
}
const LinkUITools = ({
  setAddingBlock,
  focusAddBlockButton
}) => {
  const blockInserterAriaRole = 'listbox';
  const addBlockButtonRef = (0, _element.useRef)();

  // Focus the add block button when the popover is opened.
  (0, _element.useEffect)(() => {
    if (focusAddBlockButton) {
      addBlockButtonRef.current?.focus();
    }
  }, [focusAddBlockButton]);
  return (0, _react.createElement)(_components.__experimentalVStack, {
    className: "link-ui-tools"
  }, (0, _react.createElement)(_components.Button, {
    ref: addBlockButtonRef,
    icon: _icons.plus,
    onClick: e => {
      e.preventDefault();
      setAddingBlock(true);
    },
    "aria-haspopup": blockInserterAriaRole
  }, (0, _i18n.__)('Add block')));
};
var _default = exports.default = LinkUITools;
