// @wordpress/editor/build/components/save-publish-panels/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionsPanelFill = void 0;
exports.default = SavePublishPanels;
var _react = require("react");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _element = require("@wordpress/element");
var _entitiesSavedStates = _interopRequireDefault(require("../entities-saved-states"));
var _postPublishPanel = _interopRequireDefault(require("../post-publish-panel"));
var _pluginPrePublishPanel = _interopRequireDefault(require("../plugin-pre-publish-panel"));
var _pluginPostPublishPanel = _interopRequireDefault(require("../plugin-post-publish-panel"));
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Fill,
  Slot
} = (0, _components.createSlotFill)('ActionsPanel');
const ActionsPanelFill = exports.ActionsPanelFill = Fill;
function SavePublishPanels({
  setEntitiesSavedStatesCallback,
  closeEntitiesSavedStates,
  isEntitiesSavedStatesOpen,
  forceIsDirtyPublishPanel
}) {
  const {
    closePublishSidebar,
    togglePublishSidebar
  } = (0, _data.useDispatch)(_store.store);
  const {
    publishSidebarOpened,
    hasNonPostEntityChanges
  } = (0, _data.useSelect)(select => ({
    publishSidebarOpened: select(_store.store).isPublishSidebarOpened(),
    hasNonPostEntityChanges: select(_store.store).hasNonPostEntityChanges()
  }), []);
  const openEntitiesSavedStates = (0, _element.useCallback)(() => setEntitiesSavedStatesCallback(true), []);

  // It is ok for these components to be unmounted when not in visual use.
  // We don't want more than one present at a time, decide which to render.
  let unmountableContent;
  if (publishSidebarOpened) {
    unmountableContent = (0, _react.createElement)(_postPublishPanel.default, {
      onClose: closePublishSidebar,
      forceIsDirty: forceIsDirtyPublishPanel,
      PrePublishExtension: _pluginPrePublishPanel.default.Slot,
      PostPublishExtension: _pluginPostPublishPanel.default.Slot
    });
  } else if (hasNonPostEntityChanges) {
    unmountableContent = (0, _react.createElement)("div", {
      className: "editor-layout__toggle-entities-saved-states-panel"
    }, (0, _react.createElement)(_components.Button, {
      variant: "secondary",
      className: "editor-layout__toggle-entities-saved-states-panel-button",
      onClick: openEntitiesSavedStates,
      "aria-expanded": false
    }, (0, _i18n.__)('Open save panel')));
  } else {
    unmountableContent = (0, _react.createElement)("div", {
      className: "editor-layout__toggle-publish-panel"
    }, (0, _react.createElement)(_components.Button, {
      variant: "secondary",
      className: "editor-layout__toggle-publish-panel-button",
      onClick: togglePublishSidebar,
      "aria-expanded": false
    }, (0, _i18n.__)('Open publish panel')));
  }

  // Since EntitiesSavedStates controls its own panel, we can keep it
  // always mounted to retain its own component state (such as checkboxes).
  return (0, _react.createElement)(_react.Fragment, null, isEntitiesSavedStatesOpen && (0, _react.createElement)(_entitiesSavedStates.default, {
    close: closeEntitiesSavedStates
  }), (0, _react.createElement)(Slot, {
    bubblesVirtually: true
  }), !isEntitiesSavedStatesOpen && unmountableContent);
}
