// @wordpress/editor/build/components/post-discussion/panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostDiscussionPanel;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _blockEditor = require("@wordpress/block-editor");
var _coreData = require("@wordpress/core-data");
var _store = require("../../store");
var _postTypeSupportCheck = _interopRequireDefault(require("../post-type-support-check"));
var _postComments = _interopRequireDefault(require("../post-comments"));
var _postPingbacks = _interopRequireDefault(require("../post-pingbacks"));
var _postPanelRow = _interopRequireDefault(require("../post-panel-row"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const PANEL_NAME = 'discussion-panel';
function ModalContents({
  onClose
}) {
  return (0, _react.createElement)("div", {
    className: "editor-post-discussion"
  }, (0, _react.createElement)(_blockEditor.__experimentalInspectorPopoverHeader, {
    title: (0, _i18n.__)('Discussion'),
    onClose: onClose
  }), (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 4
  }, (0, _react.createElement)(_postTypeSupportCheck.default, {
    supportKeys: "comments"
  }, (0, _react.createElement)(_postComments.default, null)), (0, _react.createElement)(_postTypeSupportCheck.default, {
    supportKeys: "trackbacks"
  }, (0, _react.createElement)(_postPingbacks.default, null))));
}
function PostDiscussionToggle({
  isOpen,
  onClick
}) {
  const {
    commentStatus,
    pingStatus,
    commentsSupported,
    trackbacksSupported
  } = (0, _data.useSelect)(select => {
    var _getEditedPostAttribu, _getEditedPostAttribu2;
    const {
      getEditedPostAttribute
    } = select(_store.store);
    const {
      getPostType
    } = select(_coreData.store);
    const postType = getPostType(getEditedPostAttribute('type'));
    return {
      commentStatus: (_getEditedPostAttribu = getEditedPostAttribute('comment_status')) !== null && _getEditedPostAttribu !== void 0 ? _getEditedPostAttribu : 'open',
      pingStatus: (_getEditedPostAttribu2 = getEditedPostAttribute('ping_status')) !== null && _getEditedPostAttribu2 !== void 0 ? _getEditedPostAttribu2 : 'open',
      commentsSupported: !!postType.supports.comments,
      trackbacksSupported: !!postType.supports.trackbacks
    };
  }, []);
  let label;
  if (commentStatus === 'open') {
    if (pingStatus === 'open') {
      label = (0, _i18n.__)('Open');
    } else {
      label = trackbacksSupported ? (0, _i18n.__)('Comments only') : (0, _i18n.__)('Open');
    }
  } else if (pingStatus === 'open') {
    label = commentsSupported ? (0, _i18n.__)('Pings only') : (0, _i18n.__)('Pings enabled');
  } else {
    label = (0, _i18n.__)('Closed');
  }
  return (0, _react.createElement)(_components.Button, {
    size: "compact",
    className: "editor-post-discussion__panel-toggle",
    variant: "tertiary",
    "aria-label": (0, _i18n.__)('Change discussion options'),
    "aria-expanded": isOpen,
    onClick: onClick
  }, (0, _react.createElement)(_components.__experimentalText, null, label));
}
function PostDiscussionPanel() {
  const {
    isEnabled
  } = (0, _data.useSelect)(select => {
    const {
      isEditorPanelEnabled
    } = select(_store.store);
    return {
      isEnabled: isEditorPanelEnabled(PANEL_NAME)
    };
  }, []);

  // Use internal state instead of a ref to make sure that the component
  // re-renders when the popover's anchor updates.
  const [popoverAnchor, setPopoverAnchor] = (0, _element.useState)(null);
  // Memoize popoverProps to avoid returning a new object every time.
  const popoverProps = (0, _element.useMemo)(() => ({
    // Anchor the popover to the middle of the entire row so that it doesn't
    // move around when the label changes.
    anchor: popoverAnchor,
    placement: 'left-start',
    offset: 36,
    shift: true
  }), [popoverAnchor]);
  if (!isEnabled) {
    return null;
  }
  return (0, _react.createElement)(_postTypeSupportCheck.default, {
    supportKeys: ['comments', 'trackbacks']
  }, (0, _react.createElement)(_postPanelRow.default, {
    label: (0, _i18n.__)('Discussion'),
    ref: setPopoverAnchor
  }, (0, _react.createElement)(_components.Dropdown, {
    popoverProps: popoverProps,
    className: "editor-post-discussion__panel-dropdown",
    contentClassName: "editor-post-discussion__panel-dialog",
    focusOnMount: true,
    renderToggle: ({
      isOpen,
      onToggle
    }) => (0, _react.createElement)(PostDiscussionToggle, {
      isOpen: isOpen,
      onClick: onToggle
    }),
    renderContent: ({
      onClose
    }) => (0, _react.createElement)(ModalContents, {
      onClose: onClose
    })
  })));
}
