// @wordpress/editor/build/components/post-last-revision/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _icons = require("@wordpress/icons");
var _url = require("@wordpress/url");
var _check = _interopRequireDefault(require("./check"));
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function LastRevision() {
  const {
    lastRevisionId,
    revisionsCount
  } = (0, _data.useSelect)(select => {
    const {
      getCurrentPostLastRevisionId,
      getCurrentPostRevisionsCount
    } = select(_store.store);
    return {
      lastRevisionId: getCurrentPostLastRevisionId(),
      revisionsCount: getCurrentPostRevisionsCount()
    };
  }, []);
  return (0, _react.createElement)(_check.default, null, (0, _react.createElement)(_components.Button, {
    href: (0, _url.addQueryArgs)('revision.php', {
      revision: lastRevisionId
    }),
    className: "editor-post-last-revision__title",
    icon: _icons.backup,
    iconPosition: "right",
    text: (0, _i18n.sprintf)( /* translators: %s: number of revisions */
    (0, _i18n.__)('Revisions (%s)'), revisionsCount)
  }));
}
var _default = exports.default = LastRevision;
