// @wordpress/editor/build-module/components/post-discussion/panel.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { Dropdown, Button, __experimentalVStack as VStack, __experimentalText as Text } from '@wordpress/components';
import { useSelect } from '@wordpress/data';
import { useState, useMemo } from '@wordpress/element';
import { __experimentalInspectorPopoverHeader as InspectorPopoverHeader } from '@wordpress/block-editor';
import { store as coreStore } from '@wordpress/core-data';

/**
 * Internal dependencies
 */
import { store as editorStore } from '../../store';
import PostTypeSupportCheck from '../post-type-support-check';
import PostComments from '../post-comments';
import PostPingbacks from '../post-pingbacks';
import PostPanelRow from '../post-panel-row';
const PANEL_NAME = 'discussion-panel';
function ModalContents({
  onClose
}) {
  return createElement("div", {
    className: "editor-post-discussion"
  }, createElement(InspectorPopoverHeader, {
    title: __('Discussion'),
    onClose: onClose
  }), createElement(VStack, {
    spacing: 4
  }, createElement(PostTypeSupportCheck, {
    supportKeys: "comments"
  }, createElement(PostComments, null)), createElement(PostTypeSupportCheck, {
    supportKeys: "trackbacks"
  }, createElement(PostPingbacks, null))));
}
function PostDiscussionToggle({
  isOpen,
  onClick
}) {
  const {
    commentStatus,
    pingStatus,
    commentsSupported,
    trackbacksSupported
  } = useSelect(select => {
    var _getEditedPostAttribu, _getEditedPostAttribu2;
    const {
      getEditedPostAttribute
    } = select(editorStore);
    const {
      getPostType
    } = select(coreStore);
    const postType = getPostType(getEditedPostAttribute('type'));
    return {
      commentStatus: (_getEditedPostAttribu = getEditedPostAttribute('comment_status')) !== null && _getEditedPostAttribu !== void 0 ? _getEditedPostAttribu : 'open',
      pingStatus: (_getEditedPostAttribu2 = getEditedPostAttribute('ping_status')) !== null && _getEditedPostAttribu2 !== void 0 ? _getEditedPostAttribu2 : 'open',
      commentsSupported: !!postType.supports.comments,
      trackbacksSupported: !!postType.supports.trackbacks
    };
  }, []);
  let label;
  if (commentStatus === 'open') {
    if (pingStatus === 'open') {
      label = __('Open');
    } else {
      label = trackbacksSupported ? __('Comments only') : __('Open');
    }
  } else if (pingStatus === 'open') {
    label = commentsSupported ? __('Pings only') : __('Pings enabled');
  } else {
    label = __('Closed');
  }
  return createElement(Button, {
    size: "compact",
    className: "editor-post-discussion__panel-toggle",
    variant: "tertiary",
    "aria-label": __('Change discussion options'),
    "aria-expanded": isOpen,
    onClick: onClick
  }, createElement(Text, null, label));
}
export default function PostDiscussionPanel() {
  const {
    isEnabled
  } = useSelect(select => {
    const {
      isEditorPanelEnabled
    } = select(editorStore);
    return {
      isEnabled: isEditorPanelEnabled(PANEL_NAME)
    };
  }, []);

  // Use internal state instead of a ref to make sure that the component
  // re-renders when the popover's anchor updates.
  const [popoverAnchor, setPopoverAnchor] = useState(null);
  // Memoize popoverProps to avoid returning a new object every time.
  const popoverProps = useMemo(() => ({
    // Anchor the popover to the middle of the entire row so that it doesn't
    // move around when the label changes.
    anchor: popoverAnchor,
    placement: 'left-start',
    offset: 36,
    shift: true
  }), [popoverAnchor]);
  if (!isEnabled) {
    return null;
  }
  return createElement(PostTypeSupportCheck, {
    supportKeys: ['comments', 'trackbacks']
  }, createElement(PostPanelRow, {
    label: __('Discussion'),
    ref: setPopoverAnchor
  }, createElement(Dropdown, {
    popoverProps: popoverProps,
    className: "editor-post-discussion__panel-dropdown",
    contentClassName: "editor-post-discussion__panel-dialog",
    focusOnMount: true,
    renderToggle: ({
      isOpen,
      onToggle
    }) => createElement(PostDiscussionToggle, {
      isOpen: isOpen,
      onClick: onToggle
    }),
    renderContent: ({
      onClose
    }) => createElement(ModalContents, {
      onClose: onClose
    })
  })));
}
