// @wordpress/editor/build-module/components/post-pingbacks/index.js
import { createElement } from "react";
/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { CheckboxControl, ExternalLink } from '@wordpress/components';
import { useDispatch, useSelect } from '@wordpress/data';

/**
 * Internal dependencies
 */
import { store as editorStore } from '../../store';
function PostPingbacks() {
  const pingStatus = useSelect(select => {
    var _select$getEditedPost;
    return (_select$getEditedPost = select(editorStore).getEditedPostAttribute('ping_status')) !== null && _select$getEditedPost !== void 0 ? _select$getEditedPost : 'open';
  }, []);
  const {
    editPost
  } = useDispatch(editorStore);
  const onTogglePingback = () => editPost({
    ping_status: pingStatus === 'open' ? 'closed' : 'open'
  });
  return createElement(CheckboxControl, {
    __nextHasNoMarginBottom: true,
    label: __('Enable pingbacks & trackbacks'),
    checked: pingStatus === 'open',
    onChange: onTogglePingback,
    help: createElement(ExternalLink, {
      href: __('https://wordpress.org/documentation/article/trackbacks-and-pingbacks/')
    }, __('Learn more about pingbacks & trackbacks'))
  });
}
export default PostPingbacks;
