// @wordpress/edit-post/build/components/welcome-guide/template.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = WelcomeGuideTemplate;
var _react = require("react");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _image = _interopRequireDefault(require("./image"));
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function WelcomeGuideTemplate() {
  const {
    toggleFeature
  } = (0, _data.useDispatch)(_store.store);
  return (0, _react.createElement)(_components.Guide, {
    className: "edit-template-welcome-guide",
    contentLabel: (0, _i18n.__)('Welcome to the template editor'),
    finishButtonText: (0, _i18n.__)('Get started'),
    onFinish: () => toggleFeature('welcomeGuideTemplate'),
    pages: [{
      image: (0, _react.createElement)(_image.default, {
        nonAnimatedSrc: "https://s.w.org/images/block-editor/welcome-template-editor.svg",
        animatedSrc: "https://s.w.org/images/block-editor/welcome-template-editor.gif"
      }),
      content: (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)("h1", {
        className: "edit-post-welcome-guide__heading"
      }, (0, _i18n.__)('Welcome to the template editor')), (0, _react.createElement)("p", {
        className: "edit-post-welcome-guide__text"
      }, (0, _i18n.__)('Templates help define the layout of the site. You can customize all aspects of your posts and pages using blocks and patterns in this editor.')))
    }]
  });
}
