// @wordpress/block-library/build/navigation/edit/navigation-menu-delete-control.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = NavigationMenuDeleteControl;
var _react = require("react");
var _components = require("@wordpress/components");
var _coreData = require("@wordpress/core-data");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
/**
 * WordPress dependencies
 */

function NavigationMenuDeleteControl({
  onDelete
}) {
  const [isConfirmDialogVisible, setIsConfirmDialogVisible] = (0, _element.useState)(false);
  const id = (0, _coreData.useEntityId)('postType', 'wp_navigation');
  const {
    deleteEntityRecord
  } = (0, _data.useDispatch)(_coreData.store);
  return (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_components.Button, {
    className: "wp-block-navigation-delete-menu-button",
    variant: "secondary",
    isDestructive: true,
    onClick: () => {
      setIsConfirmDialogVisible(true);
    }
  }, (0, _i18n.__)('Delete menu')), isConfirmDialogVisible && (0, _react.createElement)(_components.__experimentalConfirmDialog, {
    isOpen: true,
    onConfirm: () => {
      deleteEntityRecord('postType', 'wp_navigation', id, {
        force: true
      });
      onDelete();
    },
    onCancel: () => {
      setIsConfirmDialogVisible(false);
    },
    confirmButtonText: (0, _i18n.__)('Delete')
  }, (0, _i18n.__)('Are you sure you want to delete this Navigation Menu?')));
}
