// @wordpress/editor/build/components/post-switch-to-draft-button/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostSwitchToDraftButton;
var _react = require("react");
var _components = require("@wordpress/components");
var _i18n = require("@wordpress/i18n");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

// TODO: deprecate..
function PostSwitchToDraftButton() {
  const [showConfirmDialog, setShowConfirmDialog] = (0, _element.useState)(false);
  const {
    editPost,
    savePost
  } = (0, _data.useDispatch)(_store.store);
  const {
    isSaving,
    isPublished,
    isScheduled
  } = (0, _data.useSelect)(select => {
    const {
      isSavingPost,
      isCurrentPostPublished,
      isCurrentPostScheduled
    } = select(_store.store);
    return {
      isSaving: isSavingPost(),
      isPublished: isCurrentPostPublished(),
      isScheduled: isCurrentPostScheduled()
    };
  }, []);
  const isDisabled = isSaving || !isPublished && !isScheduled;
  let alertMessage;
  let confirmButtonText;
  if (isPublished) {
    alertMessage = (0, _i18n.__)('Are you sure you want to unpublish this post?');
    confirmButtonText = (0, _i18n.__)('Unpublish');
  } else if (isScheduled) {
    alertMessage = (0, _i18n.__)('Are you sure you want to unschedule this post?');
    confirmButtonText = (0, _i18n.__)('Unschedule');
  }
  const handleConfirm = () => {
    setShowConfirmDialog(false);
    editPost({
      status: 'draft'
    });
    savePost();
  };
  return (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_components.Button, {
    __next40pxDefaultSize: true,
    className: "editor-post-switch-to-draft",
    onClick: () => {
      if (!isDisabled) {
        setShowConfirmDialog(true);
      }
    },
    "aria-disabled": isDisabled,
    variant: "secondary",
    style: {
      flexGrow: '1',
      justifyContent: 'center'
    }
  }, (0, _i18n.__)('Switch to draft')), (0, _react.createElement)(_components.__experimentalConfirmDialog, {
    isOpen: showConfirmDialog,
    onConfirm: handleConfirm,
    onCancel: () => setShowConfirmDialog(false),
    confirmButtonText: confirmButtonText
  }, alertMessage));
}
