// @wordpress/editor/build/components/post-card-panel/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostCardPanel;
var _react = require("react");
var _components = require("@wordpress/components");
var _coreData = require("@wordpress/core-data");
var _data = require("@wordpress/data");
var _i18n = require("@wordpress/i18n");
var _htmlEntities = require("@wordpress/html-entities");
var _store = require("../../store");
var _constants = require("../../store/constants");
var _lockUnlock = require("../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PostCardPanel({
  actions
}) {
  const {
    title,
    icon
  } = (0, _data.useSelect)(select => {
    const {
      getEditedPostAttribute,
      getCurrentPostType,
      getCurrentPostId,
      __experimentalGetTemplateInfo
    } = select(_store.store);
    const {
      getEditedEntityRecord
    } = select(_coreData.store);
    const _type = getCurrentPostType();
    const _id = getCurrentPostId();
    const _record = getEditedEntityRecord('postType', _type, _id);
    const _templateInfo = [_constants.TEMPLATE_POST_TYPE, _constants.TEMPLATE_PART_POST_TYPE].includes(_type) && __experimentalGetTemplateInfo(_record);
    return {
      title: _templateInfo?.title || getEditedPostAttribute('title'),
      icon: (0, _lockUnlock.unlock)(select(_store.store)).getPostIcon(_type, {
        area: _record?.area
      })
    };
  }, []);
  return (0, _react.createElement)("div", {
    className: "editor-post-card-panel"
  }, (0, _react.createElement)(_components.__experimentalHStack, {
    spacing: 2,
    className: "editor-post-card-panel__header",
    align: "flex-start"
  }, (0, _react.createElement)(_components.Icon, {
    className: "editor-post-card-panel__icon",
    icon: icon
  }), (0, _react.createElement)(_components.__experimentalText, {
    numberOfLines: 2,
    truncate: true,
    className: "editor-post-card-panel__title",
    weight: 500,
    as: "h2"
  }, title ? (0, _htmlEntities.decodeEntities)(title) : (0, _i18n.__)('No Title')), actions));
}
