// @wordpress/edit-post/build/hooks/commands/use-common-commands.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useCommonCommands;
var _data = require("@wordpress/data");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _commands = require("@wordpress/commands");
var _preferences = require("@wordpress/preferences");
var _editor = require("@wordpress/editor");
var _notices = require("@wordpress/notices");
var _store = require("../../store");
var _lockUnlock = require("../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  interfaceStore
} = (0, _lockUnlock.unlock)(_editor.privateApis);
function useCommonCommands() {
  const {
    openGeneralSidebar,
    closeGeneralSidebar
  } = (0, _data.useDispatch)(_store.store);
  const {
    activeSidebar,
    isFullscreen,
    isPublishSidebarEnabled
  } = (0, _data.useSelect)(select => {
    const {
      get
    } = select(_preferences.store);
    return {
      activeSidebar: select(interfaceStore).getActiveComplementaryArea('core'),
      isPublishSidebarEnabled: select(_editor.store).isPublishSidebarEnabled(),
      isFullscreen: get('core/edit-post', 'fullscreenMode')
    };
  }, []);
  const {
    toggle
  } = (0, _data.useDispatch)(_preferences.store);
  const {
    createInfoNotice
  } = (0, _data.useDispatch)(_notices.store);
  (0, _commands.useCommand)({
    name: 'core/open-settings-sidebar',
    label: (0, _i18n.__)('Toggle settings sidebar'),
    icon: (0, _i18n.isRTL)() ? _icons.drawerLeft : _icons.drawerRight,
    callback: ({
      close
    }) => {
      close();
      if (activeSidebar === 'edit-post/document') {
        closeGeneralSidebar();
      } else {
        openGeneralSidebar('edit-post/document');
      }
    }
  });
  (0, _commands.useCommand)({
    name: 'core/open-block-inspector',
    label: (0, _i18n.__)('Toggle block inspector'),
    icon: _icons.blockDefault,
    callback: ({
      close
    }) => {
      close();
      if (activeSidebar === 'edit-post/block') {
        closeGeneralSidebar();
      } else {
        openGeneralSidebar('edit-post/block');
      }
    }
  });
  (0, _commands.useCommand)({
    name: 'core/toggle-fullscreen-mode',
    label: isFullscreen ? (0, _i18n.__)('Exit fullscreen') : (0, _i18n.__)('Enter fullscreen'),
    icon: _icons.fullscreen,
    callback: ({
      close
    }) => {
      toggle('core/edit-post', 'fullscreenMode');
      close();
      createInfoNotice(isFullscreen ? (0, _i18n.__)('Fullscreen off.') : (0, _i18n.__)('Fullscreen on.'), {
        id: 'core/edit-post/toggle-fullscreen-mode/notice',
        type: 'snackbar',
        actions: [{
          label: (0, _i18n.__)('Undo'),
          onClick: () => {
            toggle('core/edit-post', 'fullscreenMode');
          }
        }]
      });
    }
  });
  (0, _commands.useCommand)({
    name: 'core/toggle-publish-sidebar',
    label: isPublishSidebarEnabled ? (0, _i18n.__)('Disable pre-publish checks') : (0, _i18n.__)('Enable pre-publish checks'),
    icon: _icons.formatListBullets,
    callback: ({
      close
    }) => {
      close();
      toggle('core', 'isPublishSidebarEnabled');
      createInfoNotice(isPublishSidebarEnabled ? (0, _i18n.__)('Pre-publish checks disabled.') : (0, _i18n.__)('Pre-publish checks enabled.'), {
        id: 'core/editor/publish-sidebar/notice',
        type: 'snackbar'
      });
    }
  });
}
