// @wordpress/editor/build/components/template-areas/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TemplateAreas;
var _react = require("react");
var _data = require("@wordpress/data");
var _components = require("@wordpress/components");
var _blockEditor = require("@wordpress/block-editor");
var _i18n = require("@wordpress/i18n");
var _store = require("../../store");
var _lockUnlock = require("../../lock-unlock");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function TemplateAreaItem({
  area,
  clientId
}) {
  const {
    selectBlock,
    toggleBlockHighlight
  } = (0, _data.useDispatch)(_blockEditor.store);
  const templatePartArea = (0, _data.useSelect)(select => {
    const defaultAreas = select(_store.store).__experimentalGetDefaultTemplatePartAreas();
    return defaultAreas.find(defaultArea => defaultArea.area === area);
  }, [area]);
  const highlightBlock = () => toggleBlockHighlight(clientId, true);
  const cancelHighlightBlock = () => toggleBlockHighlight(clientId, false);
  return (0, _react.createElement)(_components.Button, {
    className: "editor-template-areas__item",
    icon: templatePartArea?.icon,
    onMouseOver: highlightBlock,
    onMouseLeave: cancelHighlightBlock,
    onFocus: highlightBlock,
    onBlur: cancelHighlightBlock,
    onClick: () => {
      selectBlock(clientId);
    }
  }, templatePartArea?.label);
}
function TemplateAreas() {
  const templateParts = (0, _data.useSelect)(select => (0, _lockUnlock.unlock)(select(_store.store)).getCurrentTemplateTemplateParts(), []);
  if (!templateParts.length) {
    return null;
  }
  return (0, _react.createElement)("section", {
    className: "editor-template-areas"
  }, (0, _react.createElement)(_components.__experimentalHeading, {
    level: 3,
    className: "editor-template-areas__title"
  }, (0, _i18n.__)('Areas')), (0, _react.createElement)("ul", {
    className: "editor-template-areas__list"
  }, templateParts.map(({
    templatePart,
    block
  }) => (0, _react.createElement)("li", {
    key: block.clientId
  }, (0, _react.createElement)(TemplateAreaItem, {
    area: templatePart.area,
    clientId: block.clientId
  })))));
}
