// @wordpress/edit-post/build/components/visual-editor/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var _blockEditor = require("@wordpress/block-editor");
var _header = _interopRequireDefault(require("./header"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

class VisualEditor extends _element.Component {
  constructor(props) {
    super(props);
    this.renderHeader = this.renderHeader.bind(this);
  }
  renderHeader() {
    const {
      setTitleRef
    } = this.props;
    return (0, _react.createElement)(_header.default, {
      setTitleRef: setTitleRef
    });
  }
  render() {
    const {
      safeAreaBottomInset
    } = this.props;
    return (0, _react.createElement)(_blockEditor.BlockList, {
      header: this.renderHeader,
      safeAreaBottomInset: safeAreaBottomInset
    });
  }
}
exports.default = VisualEditor;
