// @wordpress/edit-post/build/components/meta-boxes/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MetaBoxes;
var _react = require("react");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _editor = require("@wordpress/editor");
var _metaBoxesArea = _interopRequireDefault(require("./meta-boxes-area"));
var _metaBoxVisibility = _interopRequireDefault(require("./meta-box-visibility"));
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function MetaBoxes({
  location
}) {
  const registry = (0, _data.useRegistry)();
  const {
    metaBoxes,
    areMetaBoxesInitialized,
    isEditorReady
  } = (0, _data.useSelect)(select => {
    const {
      __unstableIsEditorReady
    } = select(_editor.store);
    const {
      getMetaBoxesPerLocation,
      areMetaBoxesInitialized: _areMetaBoxesInitialized
    } = select(_store.store);
    return {
      metaBoxes: getMetaBoxesPerLocation(location),
      areMetaBoxesInitialized: _areMetaBoxesInitialized(),
      isEditorReady: __unstableIsEditorReady()
    };
  }, [location]);
  const hasMetaBoxes = !!metaBoxes?.length;

  // When editor is ready, initialize postboxes (wp core script) and metabox
  // saving. This initializes all meta box locations, not just this specific
  // one.
  (0, _element.useEffect)(() => {
    if (isEditorReady && hasMetaBoxes && !areMetaBoxesInitialized) {
      registry.dispatch(_store.store).initializeMetaBoxes();
    }
  }, [isEditorReady, hasMetaBoxes, areMetaBoxesInitialized]);
  if (!areMetaBoxesInitialized) {
    return null;
  }
  return (0, _react.createElement)(_react.Fragment, null, (metaBoxes !== null && metaBoxes !== void 0 ? metaBoxes : []).map(({
    id
  }) => (0, _react.createElement)(_metaBoxVisibility.default, {
    key: id,
    id: id
  })), (0, _react.createElement)(_metaBoxesArea.default, {
    location: location
  }));
}
