// @wordpress/editor/build/components/post-taxonomies/panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _store = require("../../store");
var _index = _interopRequireDefault(require("./index"));
var _check = _interopRequireDefault(require("./check"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function TaxonomyPanel({
  taxonomy,
  children
}) {
  const slug = taxonomy?.slug;
  const panelName = slug ? `taxonomy-panel-${slug}` : '';
  const {
    isEnabled,
    isOpened
  } = (0, _data.useSelect)(select => {
    const {
      isEditorPanelEnabled,
      isEditorPanelOpened
    } = select(_store.store);
    return {
      isEnabled: slug ? isEditorPanelEnabled(panelName) : false,
      isOpened: slug ? isEditorPanelOpened(panelName) : false
    };
  }, [panelName, slug]);
  const {
    toggleEditorPanelOpened
  } = (0, _data.useDispatch)(_store.store);
  if (!isEnabled) {
    return null;
  }
  const taxonomyMenuName = taxonomy?.labels?.menu_name;
  if (!taxonomyMenuName) {
    return null;
  }
  return (0, _react.createElement)(_components.PanelBody, {
    title: taxonomyMenuName,
    opened: isOpened,
    onToggle: () => toggleEditorPanelOpened(panelName)
  }, children);
}
function PostTaxonomies() {
  return (0, _react.createElement)(_check.default, null, (0, _react.createElement)(_index.default, {
    taxonomyWrapper: (content, taxonomy) => {
      return (0, _react.createElement)(TaxonomyPanel, {
        taxonomy: taxonomy
      }, content);
    }
  }));
}
var _default = exports.default = PostTaxonomies;
