// @wordpress/editor/build/components/editor-help/view-sections.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HelpDetailSectionHeadingText = exports.HelpDetailImage = exports.HelpDetailBodyText = void 0;
var _react = require("react");
var _reactNative = require("react-native");
var _compose = require("@wordpress/compose");
var _style = _interopRequireDefault(require("./style.scss"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const HelpDetailBodyText = ({
  text
}) => {
  const bodyStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.helpDetailBody, _style.default.helpDetailBodyDark);
  return (0, _react.createElement)(_reactNative.Text, {
    selectable: true,
    style: bodyStyle
  }, text);
};
exports.HelpDetailBodyText = HelpDetailBodyText;
const HelpDetailSectionHeadingText = ({
  text,
  badge
}) => {
  const headingTextStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.helpDetailSectionHeadingText, _style.default.helpDetailSectionHeadingTextDark);
  return (0, _react.createElement)(_reactNative.View, {
    style: _style.default.helpDetailSectionHeading
  }, badge && (0, _react.createElement)(HelpDetailBadge, {
    text: badge
  }), (0, _react.createElement)(_reactNative.Text, {
    accessibilityRole: "header",
    selectable: true,
    style: headingTextStyle
  }, text));
};
exports.HelpDetailSectionHeadingText = HelpDetailSectionHeadingText;
const HelpDetailImage = ({
  accessible,
  accessibilityLabel,
  source,
  sourceDarkMode
}) => {
  const imageStyle = (0, _compose.usePreferredColorSchemeStyle)(_style.default.helpDetailImage, _style.default.helpDetailImageDark);
  const darkModeEnabled = (0, _compose.usePreferredColorScheme)() === 'dark';
  return (0, _react.createElement)(_reactNative.Image, {
    accessible: accessible,
    accessibilityLabel: accessibilityLabel,
    source: darkModeEnabled && sourceDarkMode ? sourceDarkMode : source,
    style: imageStyle
  });
};
exports.HelpDetailImage = HelpDetailImage;
const HelpDetailBadge = ({
  text
}) => {
  return (0, _react.createElement)(_reactNative.View, {
    style: _style.default.helpDetailBadgeContainer
  }, (0, _react.createElement)(_reactNative.Text, {
    style: _style.default.helpDetailBadgeText
  }, text));
};
