// @wordpress/edit-post/build/components/header/more-menu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _compose = require("@wordpress/compose");
var _editor = require("@wordpress/editor");
var _preferences = require("@wordpress/preferences");
var _keycodes = require("@wordpress/keycodes");
var _lockUnlock = require("../../../lock-unlock");
var _managePatternsMenuItem = _interopRequireDefault(require("./manage-patterns-menu-item"));
var _welcomeGuideMenuItem = _interopRequireDefault(require("./welcome-guide-menu-item"));
var _preferencesModal = _interopRequireDefault(require("../../preferences-modal"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  ToolsMoreMenuGroup,
  ViewMoreMenuGroup
} = (0, _lockUnlock.unlock)(_editor.privateApis);
const MoreMenu = () => {
  const isLargeViewport = (0, _compose.useViewportMatch)('large');
  return (0, _react.createElement)(_react.Fragment, null, isLargeViewport && (0, _react.createElement)(ViewMoreMenuGroup, null, (0, _react.createElement)(_preferences.PreferenceToggleMenuItem, {
    scope: "core/edit-post",
    name: "fullscreenMode",
    label: (0, _i18n.__)('Fullscreen mode'),
    info: (0, _i18n.__)('Show and hide the admin user interface'),
    messageActivated: (0, _i18n.__)('Fullscreen mode activated'),
    messageDeactivated: (0, _i18n.__)('Fullscreen mode deactivated'),
    shortcut: _keycodes.displayShortcut.secondary('f')
  })), (0, _react.createElement)(ToolsMoreMenuGroup, null, (0, _react.createElement)(_managePatternsMenuItem.default, null), (0, _react.createElement)(_welcomeGuideMenuItem.default, null)), (0, _react.createElement)(_preferencesModal.default, null));
};
var _default = exports.default = MoreMenu;
