// @wordpress/editor/build/components/post-pingbacks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _i18n = require("@wordpress/i18n");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _store = require("../../store");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function PostPingbacks() {
  const pingStatus = (0, _data.useSelect)(select => {
    var _select$getEditedPost;
    return (_select$getEditedPost = select(_store.store).getEditedPostAttribute('ping_status')) !== null && _select$getEditedPost !== void 0 ? _select$getEditedPost : 'open';
  }, []);
  const {
    editPost
  } = (0, _data.useDispatch)(_store.store);
  const onTogglePingback = () => editPost({
    ping_status: pingStatus === 'open' ? 'closed' : 'open'
  });
  return (0, _react.createElement)(_components.CheckboxControl, {
    __nextHasNoMarginBottom: true,
    label: (0, _i18n.__)('Enable pingbacks & trackbacks'),
    checked: pingStatus === 'open',
    onChange: onTogglePingback,
    help: (0, _react.createElement)(_components.ExternalLink, {
      href: (0, _i18n.__)('https://wordpress.org/documentation/article/trackbacks-and-pingbacks/')
    }, (0, _i18n.__)('Learn more about pingbacks & trackbacks'))
  });
}
var _default = exports.default = PostPingbacks;
