// @wordpress/editor/build/components/sidebar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _blockEditor = require("@wordpress/block-editor");
var _data = require("@wordpress/data");
var _element = require("@wordpress/element");
var _i18n = require("@wordpress/i18n");
var _icons = require("@wordpress/icons");
var _keyboardShortcuts = require("@wordpress/keyboard-shortcuts");
var _components = require("@wordpress/components");
var _interface = require("@wordpress/interface");
var _panel = _interopRequireDefault(require("../page-attributes/panel"));
var _patternOverridesPanel = _interopRequireDefault(require("../pattern-overrides-panel"));
var _pluginDocumentSettingPanel = _interopRequireDefault(require("../plugin-document-setting-panel"));
var _pluginSidebar = _interopRequireDefault(require("../plugin-sidebar"));
var _panel2 = _interopRequireDefault(require("../post-last-revision/panel"));
var _postSummary = _interopRequireDefault(require("./post-summary"));
var _panel3 = _interopRequireDefault(require("../post-taxonomies/panel"));
var _postTransformPanel = _interopRequireDefault(require("../post-transform-panel"));
var _header = _interopRequireDefault(require("./header"));
var _templateContentPanel = _interopRequireDefault(require("../template-content-panel"));
var _useAutoSwitchEditorSidebars = _interopRequireDefault(require("../provider/use-auto-switch-editor-sidebars"));
var _constants = require("./constants");
var _lockUnlock = require("../../lock-unlock");
var _store = require("../../store");
var _constants2 = require("../../store/constants");
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  Tabs
} = (0, _lockUnlock.unlock)(_components.privateApis);
const SIDEBAR_ACTIVE_BY_DEFAULT = _element.Platform.select({
  web: true,
  native: false
});
const SidebarContent = ({
  tabName,
  keyboardShortcut,
  renderingMode,
  onActionPerformed,
  extraPanels
}) => {
  const tabListRef = (0, _element.useRef)(null);
  // Because `PluginSidebar` renders a `ComplementaryArea`, we
  // need to forward the `Tabs` context so it can be passed through the
  // underlying slot/fill.
  const tabsContextValue = (0, _element.useContext)(Tabs.Context);

  // This effect addresses a race condition caused by tabbing from the last
  // block in the editor into the settings sidebar. Without this effect, the
  // selected tab and browser focus can become separated in an unexpected way
  // (e.g the "block" tab is focused, but the "post" tab is selected).
  (0, _element.useEffect)(() => {
    const tabsElements = Array.from(tabListRef.current?.querySelectorAll('[role="tab"]') || []);
    const selectedTabElement = tabsElements.find(
    // We are purposefully using a custom `data-tab-id` attribute here
    // because we don't want rely on any assumptions about `Tabs`
    // component internals.
    element => element.getAttribute('data-tab-id') === tabName);
    const activeElement = selectedTabElement?.ownerDocument.activeElement;
    const tabsHasFocus = tabsElements.some(element => {
      return activeElement && activeElement.id === element.id;
    });
    if (tabsHasFocus && selectedTabElement && selectedTabElement.id !== activeElement?.id) {
      selectedTabElement?.focus();
    }
  }, [tabName]);
  return (0, _react.createElement)(_pluginSidebar.default, {
    identifier: tabName,
    header: (0, _react.createElement)(Tabs.Context.Provider, {
      value: tabsContextValue
    }, (0, _react.createElement)(_header.default, {
      ref: tabListRef
    })),
    closeLabel: (0, _i18n.__)('Close Settings')
    // This classname is added so we can apply a corrective negative
    // margin to the panel.
    // see https://github.com/WordPress/gutenberg/pull/55360#pullrequestreview-1737671049
    ,
    className: "editor-sidebar__panel",
    headerClassName: "editor-sidebar__panel-tabs"
    /* translators: button label text should, if possible, be under 16 characters. */,
    title: (0, _i18n.__)('Settings'),
    toggleShortcut: keyboardShortcut,
    icon: (0, _i18n.isRTL)() ? _icons.drawerLeft : _icons.drawerRight,
    isActiveByDefault: SIDEBAR_ACTIVE_BY_DEFAULT
  }, (0, _react.createElement)(Tabs.Context.Provider, {
    value: tabsContextValue
  }, (0, _react.createElement)(Tabs.TabPanel, {
    tabId: _constants.sidebars.document,
    focusable: false
  }, (0, _react.createElement)(_postSummary.default, {
    onActionPerformed: onActionPerformed
  }), (0, _react.createElement)(_pluginDocumentSettingPanel.default.Slot, null), renderingMode !== 'post-only' && (0, _react.createElement)(_templateContentPanel.default, null), (0, _react.createElement)(_postTransformPanel.default, null), (0, _react.createElement)(_panel2.default, null), (0, _react.createElement)(_panel3.default, null), (0, _react.createElement)(_panel.default, null), (0, _react.createElement)(_patternOverridesPanel.default, null), extraPanels), (0, _react.createElement)(Tabs.TabPanel, {
    tabId: _constants.sidebars.block,
    focusable: false
  }, (0, _react.createElement)(_blockEditor.BlockInspector, null))));
};
const Sidebar = ({
  extraPanels,
  onActionPerformed
}) => {
  (0, _useAutoSwitchEditorSidebars.default)();
  const {
    tabName,
    keyboardShortcut,
    showSummary,
    renderingMode
  } = (0, _data.useSelect)(select => {
    const shortcut = select(_keyboardShortcuts.store).getShortcutRepresentation('core/editor/toggle-sidebar');
    const sidebar = select(_interface.store).getActiveComplementaryArea('core');
    const _isEditorSidebarOpened = [_constants.sidebars.block, _constants.sidebars.document].includes(sidebar);
    let _tabName = sidebar;
    if (!_isEditorSidebarOpened) {
      _tabName = !!select(_blockEditor.store).getBlockSelectionStart() ? _constants.sidebars.block : _constants.sidebars.document;
    }
    return {
      tabName: _tabName,
      keyboardShortcut: shortcut,
      showSummary: ![_constants2.TEMPLATE_POST_TYPE, _constants2.TEMPLATE_PART_POST_TYPE, _constants2.NAVIGATION_POST_TYPE].includes(select(_store.store).getCurrentPostType()),
      renderingMode: select(_store.store).getRenderingMode()
    };
  }, []);
  const {
    enableComplementaryArea
  } = (0, _data.useDispatch)(_interface.store);
  const onTabSelect = (0, _element.useCallback)(newSelectedTabId => {
    if (!!newSelectedTabId) {
      enableComplementaryArea('core', newSelectedTabId);
    }
  }, [enableComplementaryArea]);
  return (0, _react.createElement)(Tabs, {
    selectedTabId: tabName,
    onSelect: onTabSelect,
    selectOnMove: false
  }, (0, _react.createElement)(SidebarContent, {
    tabName: tabName,
    keyboardShortcut: keyboardShortcut,
    showSummary: showSummary,
    renderingMode: renderingMode,
    onActionPerformed: onActionPerformed,
    extraPanels: extraPanels
  }));
};
var _default = exports.default = Sidebar;
