// @wordpress/editor/build/components/sidebar/post-summary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostSummary;
var _react = require("react");
var _components = require("@wordpress/components");
var _data = require("@wordpress/data");
var _pluginPostStatusInfo = _interopRequireDefault(require("../plugin-post-status-info"));
var _postActions = _interopRequireDefault(require("../post-actions"));
var _panel = _interopRequireDefault(require("../post-author/panel"));
var _postCardPanel = _interopRequireDefault(require("../post-card-panel"));
var _postContentInformation = _interopRequireDefault(require("../post-content-information"));
var _panel2 = _interopRequireDefault(require("../post-discussion/panel"));
var _panel3 = require("../post-excerpt/panel");
var _panel4 = _interopRequireDefault(require("../post-featured-image/panel"));
var _panel5 = _interopRequireDefault(require("../post-format/panel"));
var _postLastEditedPanel = _interopRequireDefault(require("../post-last-edited-panel"));
var _postPanelSection = _interopRequireDefault(require("../post-panel-section"));
var _panel6 = _interopRequireDefault(require("../post-schedule/panel"));
var _postStatus = _interopRequireDefault(require("../post-status"));
var _postSticky = _interopRequireDefault(require("../post-sticky"));
var _postSyncStatus = _interopRequireDefault(require("../post-sync-status"));
var _panel7 = _interopRequireDefault(require("../post-template/panel"));
var _panel8 = _interopRequireDefault(require("../post-trash/panel"));
var _panel9 = _interopRequireDefault(require("../post-url/panel"));
var _store = require("../../store");
var _constants = require("../../store/constants");
var _templateAreas = _interopRequireDefault(require("../template-areas"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * Module Constants
 */
const PANEL_NAME = 'post-status';
function PostSummary({
  onActionPerformed
}) {
  const {
    isRemovedPostStatusPanel,
    postType
  } = (0, _data.useSelect)(select => {
    // We use isEditorPanelRemoved to hide the panel if it was programatically removed. We do
    // not use isEditorPanelEnabled since this panel should not be disabled through the UI.
    const {
      isEditorPanelRemoved,
      getCurrentPostType
    } = select(_store.store);
    return {
      isRemovedPostStatusPanel: isEditorPanelRemoved(PANEL_NAME),
      postType: getCurrentPostType()
    };
  }, []);
  const isPattern = postType === _constants.PATTERN_POST_TYPE;
  const isTemplate = postType === _constants.TEMPLATE_POST_TYPE;
  const isTemplatePart = postType === _constants.TEMPLATE_PART_POST_TYPE;
  const isNavigation = postType === _constants.NAVIGATION_POST_TYPE;
  return (0, _react.createElement)(_postPanelSection.default, {
    className: "editor-post-summary"
  }, (0, _react.createElement)(_pluginPostStatusInfo.default.Slot, null, fills => (0, _react.createElement)(_react.Fragment, null, (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 4
  }, (0, _react.createElement)(_postCardPanel.default, {
    actions: (0, _react.createElement)(_postActions.default, {
      onActionPerformed: onActionPerformed
    })
  }), (0, _react.createElement)(_panel4.default, {
    withPanelBody: false
  }), (0, _react.createElement)(_panel3.PrivatePostExcerptPanel, null), (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 1
  }, (0, _react.createElement)(_postContentInformation.default, null), (0, _react.createElement)(_postLastEditedPanel.default, null)), !isRemovedPostStatusPanel && (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 2
  }, (0, _react.createElement)(_components.__experimentalVStack, {
    spacing: 1
  }, (0, _react.createElement)(_postStatus.default, null), (0, _react.createElement)(_panel6.default, null), (0, _react.createElement)(_panel7.default, null), (0, _react.createElement)(_panel9.default, null), (0, _react.createElement)(_panel2.default, null), (0, _react.createElement)(_postSyncStatus.default, null)), (0, _react.createElement)(_postSticky.default, null), (0, _react.createElement)(_panel5.default, null), (0, _react.createElement)(_panel.default, null), isTemplate && (0, _react.createElement)(_templateAreas.default, null), fills, !isPattern && !isTemplate && !isTemplatePart && !isNavigation && (0, _react.createElement)(_panel8.default, null))))));
}
