// @wordpress/editor/build-module/components/editor-help/index.native.js
import { createElement, Fragment } from "react";
/**
 * External dependencies
 */
import { SafeAreaView, ScrollView, StyleSheet, View } from 'react-native';
import { TransitionPresets } from '@react-navigation/stack';

/**
 * WordPress dependencies
 */
import { BottomSheet, BottomSheetConsumer, PanelBody } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { helpFilled, plusCircleFilled, trash, cog } from '@wordpress/icons';
import { useSelect } from '@wordpress/data';
import { store as editorStore } from '@wordpress/editor';
import { requestContactCustomerSupport, requestGotoCustomerSupportOptions } from '@wordpress/react-native-bridge';

/**
 * Internal dependencies
 */
import styles from './style.scss';
import HelpDetailNavigationScreen from './help-detail-navigation-screen';
import HelpTopicRow from './help-topic-row';
import HelpGetSupportButton from './help-get-support-button';
import IntroToBlocks from './intro-to-blocks';
import AddBlocks from './add-blocks';
import MoveBlocks from './move-blocks';
import RemoveBlocks from './remove-blocks';
import CustomizeBlocks from './customize-blocks';
import moveBlocksIcon from './icon-move-blocks';
import HelpSectionTitle from './help-section-title';
const HELP_TOPICS = [{
  label: __('What is a block?'),
  slug: 'what-is-a-block',
  icon: helpFilled,
  view: createElement(IntroToBlocks, null)
}, {
  label: __('Add blocks'),
  slug: 'add-blocks',
  icon: plusCircleFilled,
  view: createElement(AddBlocks, null)
}, {
  label: __('Move blocks'),
  slug: 'move-blocks',
  icon: moveBlocksIcon,
  view: createElement(MoveBlocks, null)
}, {
  label: __('Remove blocks'),
  slug: 'remove-blocks',
  icon: trash,
  view: createElement(RemoveBlocks, null)
}, {
  label: __('Customize blocks'),
  slug: 'customize-blocks',
  icon: cog,
  view: createElement(CustomizeBlocks, null)
}];
function EditorHelpTopics({
  close,
  isVisible,
  onClose,
  showSupport
}) {
  const {
    postType
  } = useSelect(select => ({
    postType: select(editorStore).getEditedPostAttribute('type')
  }));
  const title = postType === 'page' ? __('How to edit your page') : __('How to edit your post');
  const supportSection = createElement(Fragment, null, createElement(HelpSectionTitle, null, __('Get support')), createElement(HelpGetSupportButton, {
    title: __('Contact support'),
    onPress: requestContactCustomerSupport
  }), createElement(HelpGetSupportButton, {
    title: __('More support options'),
    onPress: requestGotoCustomerSupportOptions
  }));
  return createElement(BottomSheet, {
    isVisible: isVisible,
    onClose: onClose,
    hideHeader: true,
    hasNavigation: true,
    contentStyle: styles.contentContainer,
    testID: "editor-help-modal"
  }, createElement(SafeAreaView, {
    style: styles.safeAreaContainer
  }, createElement(BottomSheet.NavigationContainer, {
    animate: true,
    main: true,
    style: styles.navigationContainer
  }, createElement(BottomSheet.NavigationScreen, {
    isScrollable: true,
    fullScreen: true,
    name: "help-topics"
  }, createElement(View, {
    style: styles.container
  }, createElement(BottomSheet.NavBar, null, createElement(BottomSheet.NavBar.DismissButton, {
    onPress: close,
    iosText: __('Close')
  }), createElement(BottomSheet.NavBar.Heading, null, title)), createElement(BottomSheetConsumer, null, ({
    listProps
  }) => {
    const contentContainerStyle = StyleSheet.flatten(listProps.contentContainerStyle);
    return createElement(ScrollView, {
      ...listProps,
      contentContainerStyle: {
        ...contentContainerStyle,
        paddingBottom: Math.max(listProps.safeAreaBottomInset, contentContainerStyle.paddingBottom),
        /**
         * Remove margin set via `hideHeader`. Combining a header
         * and navigation in this bottom sheet is at odds with the
         * current `BottomSheet` implementation.
         */
        marginTop: 0
      }
    }, createElement(PanelBody, null, createElement(HelpSectionTitle, null, __('The basics')), HELP_TOPICS.map(({
      label,
      icon,
      slug
    }, index) => {
      const isLastItem = index === HELP_TOPICS.length - 1;
      return createElement(HelpTopicRow, {
        key: slug,
        label: label,
        icon: icon,
        screenName: slug,
        isLastItem: isLastItem
      });
    }), showSupport && supportSection));
  }))), HELP_TOPICS.map(({
    view,
    label,
    slug
  }) => {
    return createElement(HelpDetailNavigationScreen, {
      key: slug,
      name: slug,
      content: view,
      label: label,
      options: {
        gestureEnabled: false,
        ...TransitionPresets.DefaultTransition
      }
    });
  }))));
}
export default EditorHelpTopics;
